Enable systemd stage 1 by default

ElvishJerricco opened the pull request:
- [x] https://github.com/NixOS/nixpkgs/pull/434312
- [x] Release notes
- [x] Improve assertions for unsupported features by pointing to releasenotes
- [x] Assert on `cryptsetup-askpass`
- [x] Double check options are unhidden (`crypttabExtraOpts` slipped through)
- [x] One of #441682 or #444829
- [x] #441777

ElvishJerricco committed:
nixos/tests/swap: Use systemd stage 1

ElvishJerricco committed:
nixos/tests/boot-stage2: Explicitly disable systemd stage 1

ElvishJerricco committed:
nixos/tests/initrd-network-ssh: Explicitly disable systemd stage 1

ElvishJerricco committed:
nixos/systemd-stage-1: Enable by default

ElvishJerricco committed:
nixos/tests/initrd-secrets: Use systemd stage 1

ElvishJerricco committed:
nixos/tests/initrd-network: Explicitly disable systemd stage 1

ElvishJerricco committed:
nixos/tests/non-default-filesystems/btrfs: Use systemd stage 1

ElvishJerricco committed:
nixos/tests/boot-stage1: Explicitly disable systemd stage 1

ElvishJerricco committed:
nixos/tests/iscsi: Explicitly disable systemd stage 1

ElvishJerricco committed:
nixos/tests/predictable-interface-names: Explicitly set 'boot.initrd.systemd.enable'

emilazy commented:
> * [ ]  [nixos/iso-image: Implement the "Ventoy Compatible" protocol #435753](https://github.com/NixOS/nixpkgs/pull/435753)

(Even if we decide to do this, it should not be a blocker as it is already quite broken. I think we should remove this from the checklist. I’ve commented on the PR.)

vcunat commented:
I created a jobset, but it wouldn't even evaluate.  See
https://hydra.nixos.org/jobset/nixos/pr-435781-systemd-stage1#tabs-errors

ElvishJerricco committed:
nixos/tests/installer: Support both stage 1 implementations

ElvishJerricco committed:
nixos/tests/ec2-nixops: Support both stage 1 implementations

ElvishJerricco committed:
nixos/tests/luks: Explicitly disable systemd stage 1

systemd stage 1 is covered by
`nixos.tests.systemd-initrd-luks-password`, but we should still test
scripted stage 1 for the time being.

ElvishJerricco committed:
nixos/tests/initrd-luks-empty-passphrase: Explicitly set 'boot.initrd.systemd.enable'

ElvishJerricco committed:
nixos/tests/qemu-vm-external-disk-image: Increase bootSize

The systemd initrd is a little larger and it's gone over the threshold.

ElvishJerricco committed:
nixos/luksroot: Unhide crypttabExtraOpts

ElvishJerricco committed:
nixos/systemd-stage-1: Enable dbus by default

Also fix the scary dbus warning when using systemctl

ElvishJerricco committed:
nixos/systemd-stage-1/users-groups: Assert against cryptsetup-askpass

ElvishJerricco committed:
nixos/tests/initrd-network-opentvpn: Explicit systemdStage1 arg

ElvishJerricco committed:
nixos/tests/hibernate: Explicit systemdStage1 arg

ElvishJerricco committed:
nixos/tests/systemd-shutdown: Explicit systemdStage1 arg

ElvishJerricco committed:
nixos/tests/envfs: Explicit systemdStage1 arg

ElvishJerricco committed:
nixos/systemd-stage-1: Support rd.systemd.break=

adamcstephens reviewed (approved):
It's been rock solid for me, for a long time, across a variety of environments. Would recommend.

griffi-gh commented on nixos/doc/manual/installation/installing-from-other-distro.section.md:
```diff
@@ -149,6 +149,10 @@ The first steps to all these are the same:
     :::
 
+    ::: {.warning}
+    The lustrate process will not work if the [](#opt-boot.initrd.systemd.enable) option is set to `true`, which is now the default. Setting this to `false` is deprecated and scheduled for removal in NixOS 26.11, along with `NIXOS_LUSTRATE`. Other installation methods, such as the one outlined above, or installing from [kexec](#sec-booting-via-kexec), are recommended instead.
```
unfortunately kexec doesnt work if the image physically cant fit into ram

NIXOS_LUSTRATE was the only way i found to get a nixos installed onto smaller OCI nodes with 1 gb of ram.


tbf theres an alternative of booting into smaller kexec image like alpine and repartitioning the drive to fit the nixos installer, but that requires a ton of manual effort for every deployed VM, and NIXOS_LUSTRATE was the only solution that just worked over ssh.

I dont think this feature should be dropped unless a viable alternative is presented.

K900 replied:
You can use the OCI image builder that is literally in nixpkgs?

griffi-gh replied:
> You can use the OCI image builder that is literally in nixpkgs?

i mean OCI As in Oracle Cloud Infrastructure

K900 replied:
That is also what I mean.

tomfitzhenry commented:
FYI, a git bisect has identified b3054f010f9d31eaf040b8961a4780d183738222 as triggering failures in nixosTests.{openssh,nfs4,samba}: https://github.com/NixOS/nixpkgs/issues/509254#issuecomment-4231589333

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574/128

Frontear commented:
Bit late but congratulations on the successful merge! I know this has been in the cooking pot for a while and I'm glad to see it finally merged in!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/systemd-stage-1-migration/77113/1

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/boot-fails-after-new-systemd-stage-1/77161/1

aszlig commented:
This looks like it broke the `system-shutdown` mechanism and thus the [`systemd` NixOS test](https://hydra.nixos.org/build/326940536).

Haven't dug into this thoroughly but in order to save others time debugging, this even fails if the test has a more robust mechanism to communicate success to the test driver via block device because I suspected it might be a race condition while filesystems are unmounted.

<details>
  <summary>Here is the diff using the block device method</summary>

```diff
diff --git a/nixos/tests/systemd.nix b/nixos/tests/systemd.nix
index 98cc88a0326f..bc1ec7760aa7 100644
--- a/nixos/tests/systemd.nix
+++ b/nixos/tests/systemd.nix
@@ -13,6 +13,7 @@
       virtualisation.emptyDiskImages = [
         512
         512
+        1 # Needed for communicating status for systemd-shutdown script.
       ];
 
       environment.systemPackages = [ pkgs.cryptsetup ];
@@ -40,18 +41,7 @@
 
       systemd.shutdown.test = pkgs.writeScript "test.shutdown" ''
         #!${pkgs.runtimeShell}
-        PATH=${
-          lib.makeBinPath (
-            with pkgs;
-            [
-              util-linux
-              coreutils
-            ]
-          )
-        }
-        mount -t 9p shared -o trans=virtio,version=9p2000.L /tmp/shared
-        touch /tmp/shared/shutdown-test
-        umount /tmp/shared
+        echo shutdown-action-done > /dev/vdd
       '';
 
       systemd.services.oncalendar-test = {
@@ -184,7 +174,10 @@
       with subtest("systemd-shutdown works"):
           machine.shutdown()
           machine.wait_for_unit("multi-user.target")
-          machine.succeed("test -e /tmp/shared/shutdown-test")
+          t.assertEqual(
+            "shutdown-action-done",
+            machine.succeed("head -c20 /dev/vdd"),
+          )
 
       # Test settings from /etc/sysctl.d/50-default.conf are applied
       with subtest("systemd sysctl settings are applied"):
```

</details>

When setting `boot.initrd.systemd.enable` to `false`, the test succeeds.

ncfavier commented:
This removed e2fsprogs from the default minimal ISO: https://github.com/NixOS/nixpkgs/issues/515555

trofi commented:
Bisect says 5e0a4d739ddd85e045e684a74660b6f8e0b9f705 `nixos/tests/installer: Support both stage 1 implementations` broke the eval of `linuxPackages.zfs_2_3.tests` in `master`:

```
$ nix-instantiate -A linuxPackages.zfs_2_3.tests
error:
       … while evaluating the attribute 'installer'
         at pkgs/os-specific/linux/zfs/2_3.nix:25:30:
           24|   // lib.optionalAttrs stdenv.isx86_64 {
           25|     inherit (nixosTests.zfs) installer;
             |                              ^
           26|   };

       … while evaluating a branch condition
         at nixos/tests/all-tests.nix:30:5:
           29|     val:
           30|     if isAttrs val then
             |     ^
           31|       if (val ? test) then

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: function 'anonymous lambda' called without required argument 'systemdStage1'
       at nixos/tests/installer.nix:1:1:
            1| {
             | ^
            2|   system ? builtins.currentSystem,
```
